pkg/logger/zerologr.go

```diff
@@ -0,0 +1,226 @@
+}
+
+// New returns a logr.Logger which is implemented by zerolog.
+func New() Zerologr {
+	return NewWithOptions(Options{})
+}
+
+// NewWithOptions returns a logr.Logger which is implemented by zerolog.
+func NewWithOptions(opts Options) Zerologr {
+	if opts.Logger == nil {
+		l := zerolog.New(os.Stdout).With().Timestamp().Logger()
+		opts.Logger = &l
+	}
+	return logger{
+		l:         opts.Logger,
+		verbosity: int(opts.Logger.GetLevel()),
+		prefix:    opts.Name,
+		values:    nil,
+	}
+}
+
+// Options that can be passed to NewWithOptions
+type Options struct {
+	// Name is an optional name of the logger
+	Name string
+	// Logger is an instance of zerolog, if nil a default logger is used
+	Logger *zerolog.Logger
+}
+
+// logger is a logr.Logger that uses zerolog to log.
+type logger struct {
+	l         *zerolog.Logger
+	verbosity int
+	prefix    string
+	values    []interface{}
+}
+
+func (l logger) Info(msg string, keysAndVals ...interface{}) {
+	if l.Enabled() {
+		var e *zerolog.Event
+		if l.verbosity < debugVerbosity {
+			e = l.l.Info()
+		} else if l.verbosity < traceVerbosity {
+			e = l.l.Debug()
+		} else {
+			e = l.l.Trace()
+		}
+		if l.prefix != "" {
+			e.Str("name", l.prefix)
+		}
+		add(e, l.values)
+		add(e, keysAndVals)
+		e.Msg(msg)
+	}
+}
+
+func (l logger) Enabled() bool {
+	var lvl zerolog.Level
+	if l.verbosity < debugVerbosity {
+		lvl = zerolog.InfoLevel
+	} else if l.verbosity < traceVerbosity {
+		lvl = zerolog.DebugLevel
+	} else {
+		lvl = zerolog.TraceLevel
+	}
+	if lvl < zerolog.GlobalLevel() {
+		return false
+	}
+	return true
+}
+
+func (l logger) Error(err error, msg string, keysAndVals ...interface{}) {
+	e := l.l.Error().Err(err)
+	if l.prefix != "" {
+		e.Str("name", l.prefix)
+	}
+	add(e, l.values)
+	add(e, keysAndVals)
+	e.Msg(msg)
+}
+
+func (l logger) V(verbosity int) logr.InfoLogger {
+	new := l.clone()
+	new.verbosity = verbosity
+	return new
+}
+
+// WithName returns a new logr.Logger with the specified name appended. zerologr
+// uses '/' characters to separate name elements.  Callers should not pass '/'
+// in the provided name string, but this library does not actually enforce that.
+func (l logger) WithName(name string) logr.Logger {
+	new := l.clone()
+	if len(l.prefix) > 0 {
+		new.prefix = l.prefix + "/"
+	}
+	new.prefix += name
+	return new
+}
+func (l logger) WithValues(kvList ...interface{}) logr.Logger {
+	new := l.clone()
+	new.values = append(new.values, kvList...)
+	return new
+}
+
+// Infof logs a formatted info level log to the console
+func (l logger) Infof(format string, v ...interface{}) {
+	l.Info(fmt.Sprintf(format, v...))
+}
+
+func Infof(format string, v ...interface{}) {
+	mu.RLock()
+	defer mu.RUnlock()
+	log.Info(fmt.Sprintf(format, v...))
+}
+
+// Errorf logs a formatted error level log to the console
+func (l logger) Errorf(format string, v ...interface{}) {
+	l.Error(nil, fmt.Sprintf(format, v...))
+}
+
+func Errorf(format string, v ...interface{}) {
+	mu.RLock()
```

Comment:
We don't need any locking in this package.  Initialization code need not be thread-safe and logging objects are already safe.